### PR TITLE
[RFC] Version output cleanup

### PR DESCRIPTION
**Pull request by ghost**
Discussed here https://github.com/neovim/neovim/pull/1849#discussion_r23252398

The first commit was done because some #ifdefs were in the way in `version.c` (`*_BUILTIN_TCAPS`).

### CODE REVIEW

File: src/nvim/version.c [hunk 1]
```diff
@@ -43,54 +43,17 @@ char *version_cflags = "Compilation: " NVIM_VERSION_CFLAGS;
...
-  "+extra_search",
-  "+farsi",
-  "+file_in_path",
-  "+find_in_path",
-  "+folding",
-  "-footer",
 
   // only interesting on Unix systems
 #if defined(UNIX)
   "+fork()",
```
Comment (Pyrohh) [inline src/nvim/version.c]: Does this need to be kept?

**Comment by jszakmeister** [comment]
I'm on the fence about whether this is a good idea or not.  I could go either way.

**Comment by Pyrohh** [comment]
I'm not sure if you saw the commit message, but I'll expand on it. IMHO, the array of features being printed is currently close to useless. Yes, people can get an idea of what's present in Neovim and what isn't, but that assumes they're aware of how our code has been made more monolithic. Someone who thinks our code is more modularized than it actually is (i.e., still contains ifdef'd features) could see a `-` next to a feature and assume that it could be compiled in, so they might waste their time attempting to recompile before realizing the feature doesn't exist or is already built in. More so, all of the constant feature flags drown out the few flags which are actually optional at compile time.

I have a PR documenting nvim's differences from vim, which I think would be a much better place to document such things.

**Comment by jszakmeister** [comment]
> I'm not sure if you saw the commit message, but I'll expand on it. IMHO, the array of features being printed is currently close to useless. Yes, people can get an idea of what's present in Neovim and what isn't, but that assumes they're aware of how our code has been made more monolithic. Someone who thinks our code is more modularized than it actually is (i.e., still contains ifdef'd features) could see a `-` next to a feature and assume that it could be compiled in, so they might waste their time attempting to recompile before realizing the feature doesn't exist or is already built in. More so, all of the constant feature flags drown out the few flags which are actually optional at compile time.

I definitely see that side of it.  On the other hand, someone familiar with Vim can look at it say "this is the feature set" or perhaps "ooh, they should really bring back support for that feature".  That's why I'm on the fence.  It doesn't mean I'm voting it down, only that I don't know which direction to vote in. :-)

Outside of that, the actual changes look good.  

> I have a PR documenting nvim's differences from vim, which I think would be a much better place to document such things.

That depends on where it's at--documentation can be hard to dig through. :-)

**Comment by elmart** [comment]
I also see both sides, but I'm more on the direction of this PR. A well-known feature of neovim is the removal of lots conditional compilation cruft, and the inclusion of the "huge" feature set by default.
For the experienced vim user, the inclusion of features in version string can be confusing, leading them to think they can add/remove features.
For the non-experienced user, they won't miss anything.
Now, as it's said, it should be clearly documented which of previously optional features are included, and which aren't. I would include that info in a `:help` page within nvim.

**Comment by Pyrohh** [comment]
(I would quote you, but I'm on a mobile email client, so...)

I see what you mean, especially with the "ooh this feature should be brought back". Still, that can adequately be done in the vim-differences doc I mentioned, which is the first section in`:h nvim`. Note how I said _can adequately_, as I haven't actually documented the "inlined" features yet.

**Comment by jszakmeister** [comment]
> For the experienced vim user, the inclusion of features in version string can be confusing, leading them to think they can add/remove features.

I see having to dig either way: to figure out whether Neovim supports a given feature, or whether the build will allow me to optionally turn on a feature.  Either way time is spent trying to figure it out. :-)

As I said before, I'm neutral.  Given that, as long as @elmart think the changes are good, then he should merge them.  There seems to be a consensus that it's a Good Thing. :-)

**Comment by elmart** [comment]
Given that this is a front-and-visible change, I'd like to have a broader agreement before merging. @justinmk? @tarruda?

File: src/nvim/version.c (hunk 1 above)
Comment (fwalch) [inline src/nvim/version.c]: Probably not, if the other stuff is removed.

**Comment by fwalch** [comment]
> Now, as it's said, it should be clearly documented which of previously optional features are included, and which aren't. I would include that info in a :help page within nvim.

That sounds like a good idea. Also, what about adding text like the following to `--version` to replace the list of compiled-in features:

```
For differences from Vim, see :help vim-differences.
For included Vim features, see :help nvim-vim-features.
```

The last line refers to these "compiled" features; I'm sure the wording can be improved.

**Comment by elmart** [comment]
> what about adding text like [...]

:+1:

**Comment by jszakmeister** [comment]
> That sounds like a good idea. Also, what about adding text like the following to --version to replace the list of compiled-in features:
> 
> For differences from Vim, see :help vim-differences.
> For included Vim features, see :help nvim-vim-features.

A definite +1.

**Comment by justinmk** [comment]
:+1:

File: src/nvim/version.c (hunk 1 above)
Comment (Pyrohh) [inline src/nvim/version.c]: I think it's be redundant, as it looks like `f_system` was reimplemented on top of libuv, and how libuv does things is generally out of the scope of our docs. I'll just remove it.

**Comment by Pyrohh** [comment]
Repushed, how's this?

```
...

Compiled by michael@orchid

Features included (+) or not (-):  +acl   +iconv
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/usr/local/share/nvim"
```

I figured the features can all be on one line, at least for now given how few things are optional at compile-time. If we're doing this then #2141 should be merged first (I'll get to updating that soon).

**Comment by elmart** [comment]
`Optional features included (+) or not (-)`
may be better? (added 'Optional')

**Comment by justinmk** [comment]
Compile-time features should not be an end-user concern. The goal is to remove the need for it completely. Preserving it in the `:version` dialog serves no purpose.

**Comment by Pyrohh** [comment]
@justinmk are you responding to @elmart or me? Also I don't really understand what you mean.

I agree that compile time features should not be an end user concern, but are you suggesting that we don't mention whether nvim has been built with ACL or iconv support? I ask because we still have many `HAVE_(ACL|ICONV)` ifdefs, which I assume aren't necessarily defined, so just not mentioning that may cause confusion in bug reports.

**Comment by justinmk** [comment]
Ok, I hope it's temporary though.

**Comment by Pyrohh** [comment]
> Optional features included (+) or not (-)

:+1: I just repushed with that.

**Comment by Pyrohh** [comment]
Thanks @justinmk

**Pull request merged**